README.md
# Lemon

**x86 32-bit minimal operating system**  
*Copyright (C) Adrien RICCIARDI*

## Help

* Type `help` to the Lemon prompt shell to get help, commands list and information about system usage.
* Whole system documentation can be found in repository [Documentation](https://github.com/RICCIARDI-Adrien/Lemon/tree/master/Documentation) directory.


Documentation/Change_Log.txt
10/08/2013 : remplacement des parties de la libc écrites à la main par NewLib

21/08/2013 :
- rétablissement de la libc custom, NewLib est impossible à maintenir
- ajout du symbole _end au linker script des applications. Ce symbole permet de connaître la fin du programme depuis le binaire (utile pour malloc())

22/08/2013 :
- changement du point d'entrée et de l'adresse de chargement des programmes userspace : maintenant 256 octets sont réservés aux arguments de la ligne de commande et le tas est situé après le code
- réécriture de KernelStartProgram() en assembleur "gcc"
- réécriture optimisée de la fonction SystemCall() de Libraries (évite de copier eax dans une variable et de retourner la variable)
- suppression du type bool du noyau (inutile)
- ajout fichier Strings.h pour plus de cohérence
- réécriture ShellCommandRenameFile() pour fournir plus de messages d'erreur

23/08/2013 :
- ajout du syscall SYSTEM_CALL_GET_USER_MEMORY_SIZE pour permettre à malloc() d'ajuster dynamiquement la taille du tas au démarrage d'un programme
- implémentation du tas situé après les données du programme
- début modification des arguments de la ligne de commande passés par le shell à un programme

24/08/2013 :
- suite et fin modif arguments ligne de commande

08/09/2013 :
- putchar() et puts() sont maintenant 100% conformes à POSIX 1003.1
- ajout de l'appel système FileRead() qui remplace FileReadNextBlock(), ce qui simplifie énormément le code du noyau, des bibliothèques, des appels système et des applications userspace
- modif du type FILE dans la libc pour stocker uniquement l'index du descripteur de fichier
- modif de fopen(), fread() et fclose() pour gérer le nouveau syscall et le nouveau type FILE
- ajout du support des fichiers à fgets()
- ajout d'une version minimale de fscanf() supportant seulement %d
- ajout des fonctions isspace() et isdigit() de la bibliothèque ctype (et ajout du header de la bibliothèque)

09/09/2013 :
- fin de l'implémentation de "%d" dans fscanf()

10/09/2013 :
- début de l'ajout du syscall FileWrite()
- modification de FileOpen() et FileClose() pour intégrer FileWrite()
- ajout du test de checksum pour le syscall FileRead() parmi les applications

11/09/2013 :
- amélioration de FileLoad() :
  # évite de charger un fichier de taille supérieure à la RAM
  # ne fait rien lorsqu'un fichier a une taille nulle
  # si demandé, charge le premier bloc du fichier et teste la présence du magic number des programmes avant de continuer à charger le fichier (si le magic number est absent, une erreur est retournée et le fichier n'est pas chargé) => plus besoin de charger un fichier en entier pour déterminer si c'est un programme ou non
- amélioration de la fonction de lancement de programme du shell pour éviter de lancer le contenu précédent de la mémoire lorsqu'un fichier vide est chargé
- suppression des appels système FileLoad() et FileCreate(), redondants avec FileOpen(), FileClose(), FileRead() et FileWrite()
- l'appel d'un syscall inexistant renvoie -1
- la commande de téléchargement du shell ne télécharge plus de fichiers vides ni de fichiers de taille supérieure à la RAM userspace (modification du serveur pour implémenter ce comportement)

20/09/2013 :
- changement du code de chargement du TSS en code assembleur GCC

28/09/2013 :
- sauvegarde de la version 2.0.4 du système, passage à la version 2.1.0 (ci-dessous les principaux changements à réaliser)
  # makefile dans le dossier racine qui permet de compiler facilement le système (make clean suivi de make) => suppression du script Create_Release.sh
  # le nouveau makefile permet de sélectionner facilement la langue de compilation de tout le système
  # nouvel installateur plus simple à maintenir et entièrement réécrit pour se débarrasser de l'installateur actuel rapidement bricolé
    + le MBR sera capable de charger plus de 64 Ko en mémoire (mais pas plus d'1 Mo car 16 bits), l'installateur se chargera à l'adresse 10000h pour éviter les problèmes d'alignement
  # rajout du support du timer 2 (fréquence programmable) pour disposer d'un timer et accessoirement activer le haut parleur du PC

29/09/2013 :
- amélioration du makefile du système pour éviter de tout recompiler à chaque fois (rajout du dossier Objects, création de dépendances propres, rajout du support des langues à la compilation)
- déplacement des chaînes contenues dans la commande "help" du shell dans "Strings.h"
- rajout du support des langues dans les Libraries (même si c'est inutilisé pour l'instant)
- création du makefile principal destiné à remplacer Create_Release.sh

12/12/2013 :
- passage à la version 2.1.1 du système, première fois que le terme Minimal Implementation Architecture (MiA) est apparu dans les sources (précisément dans l'en-tête doxygen de Kernel.h)
- changement des commentaires doxygen qui se trouvent en en-tête des fichiers .c pour que la documentation du module ne se trouve plus que dans les .h (simplifie la maintenance)
- suppression de Console.c et Console.h, création de Screen.c et Screen.h qui gèrent désormais toutes les fonctions liées à l'affichage. Le module Keyboard contient maintenant la fonction KeyboardReadString()
- nettoyage des fonctions liées à l'ancien module Console
- rajout du symbole DOXYGEN dans le doxyfile du système (permet d'utiliser #ifdef DOXYGEN dans les sources pour modifier la documentation générée par Doxygen)
- ajout d'un appel système permettant de copier un buffer représentant l'écran dans la mémoire vidéo en attendant le signal VSync dans le but d'éviter l'apparition d'artefacts désagréables sur l'écran (permettra de gérer une interface "graphique" depuis l'userspace et de faire des jeux)

13/12/2013 :
- System : renommage des constantes représentant les appels systèmes dans le noyau pour plus de clarté et réorganisation dans System_Calls.c/h
- System : renommage de KeyboardReadKey() en KeyboardReadChar() pour que les API soient unifiées

System/Includes/Kernel.h
/** @file Kernel.h
 * Lemon 32-bit kernel.
 * @author Adrien RICCIARDI
 * @version 1.0 : 15/06/2011
 * @version 1.1 : 16/11/2011
 * @version 1.2 : 27/04/2012
 * @version 2.0 : 27/07/2012
 * @version 2.0.1 : 12/09/2012 (protection enabled 06/10/2012).
 * @version 2.0.2 : 22/12/2012, started shell implementation.
 * @version 2.0.3 : 29/12/2012, added shell to kernel in order to keep system simple.
 * @version 2.0.4 : 13/07/2013, global code optimization.
 * @version 2.1.0 : 28/09/2013, started Minimal Implementation Architecture improvements on the whole system.
 * @version 2.1.1 : 12/12/2013, removed Console module and created Screen module where drawing functions are now located, keyboard input functions previously located in Console are now in Keyboard.
 * @version 2.1.2 : 26/04/2014, added Includes directory, global source code reorganization and improvements.
 * @version 2.1.3 : 13/05/2014, cleaned and optimized memory protection code and interrupt code.
 * @version 2.1.4 : 07/06/2014, the system is now able to compile on 64-bit development system, file system is terminated.
 * @version 2.1.5 : 26/10/2014, removed 'help' command and added 'version' command.
 */
#ifndef H_KERNEL_H
#define H_KERNEL_H

#include <Configuration.h>

//-------------------------------------------------------------------------------------------------
// Types
//-------------------------------------------------------------------------------------------------
#define NULL ((void *) 0)

//-------------------------------------------------------------------------------------------------
// Constants
//-------------------------------------------------------------------------------------------------
// Kernel memory areas
/** The MBR stores file system informations. */
#define KERNEL_MBR_ADDRESS 0x7C00
/** Kernel stack address. */
#define KERNEL_STACK_ADDRESS 0x10000

// Userspace and user programs stuff
/** Amount of RAM allowed to user space. */
#define KERNEL_USER_SPACE_SIZE ((CONFIGURATION_SYSTEM_TOTAL_RAM_SIZE_MEGA_BYTES - 1) * 1024 * 1024)
/** User space base address. This reserves the first MB of RAM for the kernel. */
#define KERNEL_USER_SPACE_ADDRESS 0x100000
/** Instruction pointer start address for user program. */
#define KERNEL_PROGRAM_ENTRY_POINT 0x100
/** User program will be loaded at this address. 256 bytes are reserved to store command line arguments. */
#define KERNEL_PROGRAM_LOAD_ADDRESS (KERNEL_USER_SPACE_ADDRESS + KERNEL_PROGRAM_ENTRY_POINT)
/** Executable program magic number. */
#define KERNEL_PROGRAM_MAGIC_NUMBER 0x83E58955 // This is the hexadecimal representation of the main() stack frame (push ebp...)

// Functions return codes
/** No error happened. */
#define ERROR_CODE_NO_ERROR 0
/** The File List is full. */
#define ERROR_CODE_FL_FULL 1
/** There are no more free block in the Blocks Allocation Table. */
#define ERROR_CODE_BAT_FULL 2
/** The provided file name is not valid. */
#define ERROR_CODE_BAD_FILE_NAME 3
/** The requested file doesn't exist in the file system. */
#define ERROR_CODE_FILE_NOT_FOUND 4
/** The file is already existing. */
#define ERROR_CODE_FILE_ALREADY_EXISTS 5
/** The maximum opened files count is reached. */
#define ERROR_CODE_CANT_OPEN_MORE_FILES 6
/** The file is opened yet. */
#define ERROR_CODE_FILE_OPENED_YET 7
/** A file was opened with an invalid mode. */
#define ERROR_CODE_UNKNOWN_OPENING_MODE 8
/** An attempt to read or write a file which was not previously opened was done. */
#define ERROR_CODE_FILE_NOT_OPENED 9
/** A file opened in read mode is accessed in write mode or a file opened in write mode is accessed in read mode. */
#define ERROR_CODE_BAD_OPENING_MODE 10
/** A bad file descriptor (value >= CONFIGURATION_FILE_SYSTEM_MAXIMUM_OPENED_FILES_COUNT) was provided to a file function. */
#define ERROR_CODE_BAD_FILE_DESCRIPTOR 11
/** Bad parameters were provided to UART during initialization. */
#define ERROR_CODE_BAD_UART_PARAMETERS 12
/** There is not enough room in RAM to load the file. */
#define ERROR_CODE_FILE_LARGER_THAN_RAM 13
/** The file is not an executable program. */
#define ERROR_CODE_FILE_NOT_EXECUTABLE 14

//-------------------------------------------------------------------------------------------------
// Functions
//-------------------------------------------------------------------------------------------------
/** Clear the kernel stack and jump to shell. */
void KernelStartShell(void);

/** Switch to protection ring 3 and execute user program's entry point. */
void KernelStartProgram(void);

#endif